### PR TITLE
WARN,TST check stacklevel for all warnings 

### LINKED ISSUE #47919: WARN,TST check stacklevel for all warnings
Currently, the stacklevel is only checked for two classes of warnings:

https://github.com/pandas-dev/pandas/blob/14de3fd9ca4178bfce5dd681fa5d0925e057c04d/pandas/_testing/_warnings.py#L132-L134

it would be good to extend this to other (all?) classes of warnings, and fixing the parts of the codebase where this fails

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #47919 (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/plotting/test_misc.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def test_scatter_matrix_axis(self, pass_axis):
             df = DataFrame(np.random.randn(100, 3))
 
         # we are plotting multiples on a sub-plot
-        with tm.assert_produces_warning(UserWarning, raise_on_extra_warnings=True):
```
Comment (MarcoGorelli) [inline pandas/tests/plotting/test_misc.py:110]: `raise_on_extra_warnings=True` is the default, so removing it

File: pandas/tests/plotting/test_hist_method.py [hunk 2]
```diff
@@ -659,13 +659,13 @@ def test_grouped_hist_layout(self, hist_df):
         with pytest.raises(ValueError, match=msg):
             df.hist(column="height", by=df.category, layout=(-1, -1))
 
-        with tm.assert_produces_warning(UserWarning):
+        with tm.assert_produces_warning(UserWarning, check_stacklevel=False):
             axes = _check_plot_works(
                 df.hist, column="height", by=df.gender, layout=(2, 1)
             )
```
Comment (MarcoGorelli) [inline pandas/tests/plotting/test_hist_method.py:665]: Separately it'd be good to have `_check_plot_works_check_warnings` (analogous to `parser.read_csv_check_warnings`), but for now I'm just ignoring the stacklevel check

**Review by mroeschke — COMMENTED** [review]
Do you think it's feasible have a precommit check enforce that `stacklevel=find_stack_level(inspect.currentframe())` is always set?

**Comment by MarcoGorelli** [comment]
Should be possible eventually, but at the moment there are some places where it needs setting manually for it to work, e.g.

https://github.com/pandas-dev/pandas/blob/a6aaeb6baf679fe133e968e0f65199fc56d177b2/pandas/compat/pickle_compat.py#L108

It looks like there, the going back up the frames temporarily brings us into the standard library:
```python
(Pdb)  frame
<frame at 0x7f87356b1b20, file '/home/marco/pandas-marco/pandas/compat/pickle_compat.py', line 92, code __new__>
(Pdb) p frame.f_back
<frame at 0x7f873569f7c0, file '/home/marco/pandas-marco/pandas/compat/pickle_compat.py', line 231, code load_newobj>
(Pdb) p frame.f_back.f_back
<frame at 0x7f87356b3640, file '/home/marco/miniconda3/envs/pandas-dev/lib/python3.8/pickle.py', line 1212, code load>
(Pdb) p frame.f_back.f_back
<frame at 0x7f87356b3640, file '/home/marco/miniconda3/envs/pandas-dev/lib/python3.8/pickle.py', line 1212, code load>
(Pdb) p frame.f_back.f_back.f_back
<frame at 0x7f8735765800, file '/home/marco/pandas-marco/pandas/compat/pickle_compat.py', line 277, code load>
(Pdb) p frame.f_back.f_back.f_back.f_back
<frame at 0x7f87356a6c40, file '/home/marco/pandas-marco/pandas/io/pickle.py', line 213, code read_pickle>
(Pdb) p frame.f_back.f_back.f_back.f_back.f_back
<frame at 0x7f873569f400, file '/home/marco/pandas-marco/pandas/tests/io/test_pickle.py', line 240, code test_legacy_sparse_warning>
```
, I'll have a look at what can be done about that

Any places where the stacklevel isn't correct should now be caught by the tests anyway, as with this PR it'd be checked for all warnings and `check_stacklevel=True` is the default

**Comment by mroeschke** [comment]
Thanks for confirming. In any case, a check can be made in a follow up

**Review by mroeschke — APPROVED** [review]

**Pull request merged**